### ISSUE TITLE
diagnostic signs disappear when entering insert mode

### ISSUE THREAD

**Issue by kaddkaka** [neovim/neovim#26078]
### Problem

LSP diagnostic signs disappear one second after entering insert mode, only to reappear after leaving insert mode.

This leads thel code "jumping" when insert mode  is entered due to the sign column disappearing. 

This is the closest thing I could find in the help (the default seems to be what I want, but the behavior is unwanted since diagnostics *are* updated):

    :helpgrep update_in_insert
    update_in_insert: (default false) Update diagnostics in
    Insert mode (if false, diagnostics are updated on
    InsertLeave)

Similar issues:

1. [https://www.reddit.com/r/neovim/comments/10z4q16/lsp\_diagnostics\_disappear\_completely\_in\_insert/](https://www.reddit.com/r/neovim/comments/10z4q16/lsp_diagnostics_disappear_completely_in_insert/)
2. [https://vi.stackexchange.com/questions/42126/neovim-how-to-keep-the-diagnostic-message-in-insert-mode](https://vi.stackexchange.com/questions/42126/neovim-how-to-keep-the-diagnostic-message-in-insert-mode)

### Expected behavior

- No diagnostics should disappear when entering insert mode.
  - (All diagnostics that were shown before entering insert mode should stay visible until insert mode is leaved)
- No new diagnostics should appear while in insert mode.

**Comment by mfussenegger** [comment]
There is afaik nothing in Neovim that would change the diagnostics on insert enter by default.

I also can't reproduce it:

![recording](https://github.com/neovim/neovim/assets/38700/e8cb673e-db3a-4900-9b17-658e3da12446)


My best guess is that if you enter insert mode and type something, the client notifies the server about the document changes and then the server re-publishes the diagnostics.

Would need some minimal reproduction steps

**Comment by liskin** [comment]
> There is afaik nothing in Neovim that would change the diagnostics on insert enter by default.

Indeed, if by "insert enter" you mean <kbd>i</kbd>, but <kbd>o</kbd> does trigger this issue reliably in LSPs that publish diagnostics before the file is saved.

> My best guess is that if you enter insert mode and type something, the client notifies the server about the document changes and then the server re-publishes the diagnostics.

Yes, that's exactly what happens. The problem is that `lua.vim.diagnostic.show` unconditionally (even if `update_in_insert` is `false`) hides the diagnostics:
https://github.com/neovim/neovim/blob/50284d07b6f020c819aeb07bfb30d88453e63b6d/runtime/lua/vim/diagnostic.lua#L1464
and then schedules them to be shown after insert mode ends:
https://github.com/neovim/neovim/blob/50284d07b6f020c819aeb07bfb30d88453e63b6d/runtime/lua/vim/diagnostic.lua#L1477-L1481

The hide should be scheduled as well, not performed immediately.

> Would need some minimal reproduction steps

1. use an LSP that publishes diagnostics while typing (as opposed to only when the file is saved), such as lua_ls
2. set `update_in_insert` to false
3. open a file that has at least one diagnostic somewhere
4. press <kbd>o</kbd>

---

Note: I was quite annoyed by this issue, but even more annoying was the fact that any normal-mode edits (deleting a word, …) would spam me with diagnostics immediately, so locally I have a miniplugin that delays hiding/showing diagnostics until after the file is saved. It happens to work around this bug. Code here: https://github.com/liskin/dotfiles/blob/2e674efcf45140d911ac28fec87dddbb568cf47e/.config/nvim/lua/init/diagnostic_delay.lua
It'd be nice if neovim itself had an option to never update diagnostics until after the file is saved. I should probably open a separate issue for this though.

**Comment by liskin** [comment]
@wookayin can you pls drop the needs: labels so the bot keeps it open?

**Comment by kaddkaka** [comment]
As a workaround that I think I will keep as my default setting because I started liking it, I added this to my config:

```vim
lua vim.diagnostic.config({virtual_text={format=function(d) return "" end}, signs=false})
nnoremap <c-j> <cmd>lua vim.diagnostic.goto_next({float={source=true}})<cr>
nnoremap <c-k> <cmd>lua vim.diagnostic.goto_prev({float={source=true}})<cr>
```

**Comment by OnlyWick** [comment]
Same issue here. The same configuration file, there will be problems on my Mac, and it will run normally on my Windows.

Mac version: macOs Sonoma 14.3.1

**I'm going crazy.**

**Comment by OnlyWick** [comment]
Following code can solve this issue.
```
vim.diagnostic.config({
    virtual_text = true,
    signs = true,
    update_in_insert = true,  // here
    underline = true,
    severity_sort = false,
    float = true,
  })
```

**Comment by kaddkaka** [comment]
> Following code can solve this issue.
> 
> ```
> vim.diagnostic.config({
>     virtual_text = true,
>     signs = true,
>     update_in_insert = true,  // here
>     underline = true,
>     severity_sort = false,
>     float = true,
>   })
> ```

This is further away from the expected behavior. This issue is about diagnostics *not* updating in insert mode.

**Comment by tkr-sh** [comment]
https://github.com/neovim/neovim/issues/26078#issuecomment-2027413996

This solved the issue for me. Thanks!

**Comment by subtleseeker** [comment]
Today is a 1-year anniversary of this issue. Any updates on this issue?

**Comment by justinmk** [comment]
> The problem is that `lua.vim.diagnostic.show` unconditionally (even if `update_in_insert` is `false`) hides the diagnostics:
> 
> [neovim/runtime/lua/vim/diagnostic.lua](https://github.com/neovim/neovim/blob/50284d07b6f020c819aeb07bfb30d88453e63b6d/runtime/lua/vim/diagnostic.lua#L1464)

That will lead to "stale" diagnostic signs in insert-mode. Which would be reported as a different bug...

**Comment by liskin** [comment]
> > The problem is that `lua.vim.diagnostic.show` unconditionally (even if `update_in_insert` is `false`) hides the diagnostics:
> > [neovim/runtime/lua/vim/diagnostic.lua](https://github.com/neovim/neovim/blob/50284d07b6f020c819aeb07bfb30d88453e63b6d/runtime/lua/vim/diagnostic.lua#L1464)
> 
> That will lead to "stale" diagnostic signs in insert-mode. Which would be reported as a different bug...

No, it wouldn't. What happens is that you enter insert mode, do something, the LSP sends new diagnostics, `lua.vim.diagnostic.show` unconditionally calls `hide`, thus making the diagnostic signs disappear, and then since `update_in_insert` is false, it doesn't show the updated diagnostics.

**Comment by liskin** [comment]
Oh, you mean it would lead to stale diagnostics if this was fixed by not hiding them? Yeah I suppose it would. But then `update_in_insert` might as well be called `hide_in_insert` :-D

**Comment by justinmk** [comment]
> setting `update_in_insert` to false kinda implies that stale diagnostics is what one wants

ok, send a PR :)

**Comment by arkriny** [comment]
I find content shifting annoying not only in insert mode, but also for other edits like deleting, pasting or commenting. Some LSP servers refresh diagnostics on save or after code action edits, which also causes "jumping". Basically users have choice to avoid this by setting 'signcolumn' or by disabling `diagnostic.config.signs`, depending on whether they want signs always or never.

I propose making the default for the signs option match 'signcolumn' (i.e., disabled by default) and documenting 'signcolumn' as the way to enable signs. Diagnostics are already clearly indicated by highlights, and the recently added statusline counter provides additional visibility. I'm not in favor of making the "stale" diagnostics behavior the default (#38313). If a choice must be made, having up-to-date diagnostics on screen is more important than having a signs time to time. Also this change would make it feasible to enable `update_in_insert` by default, since with the current defaults it causes a disaster on typing.

**Comment by justinmk** [comment]
> this change would make it feasible to enable `update_in_insert` by default, since with the current defaults it causes a disaster on typing.

Does that mean anyone that opted-in to signs would still have that disaster (or they have to disable `update_in_insert`)?

**Comment by mfussenegger** [comment]
>  Basically users have choice to avoid this by setting 'signcolumn' or by disabling diagnostic.config.signs, depending on whether they want signs always or never.

> I propose making the default for the signs option match 'signcolumn' (i.e., disabled by default) and documenting 'signcolumn' as the way to enable signs. 

Could also consider changing the sign text to an empty string and only have a `numhl` with Error -> ErrorMsg, Warn -> WarningMsg, Info/Hint -> Underlined

More compact, avoids layout shifts and still gives an indication.

**Comment by arkriny** [comment]
> Does that mean anyone that opted-in to signs would still have that disaster (or they have to disable `update_in_insert`)?

If they opt in the "right way" by setting 'signcolumn' to 'yes', they wouldn't experience it. However, they could still enable only `diagnostic.config.signs` and get the current default behavior. Most users who set 'signcolumn' beforehand wouldn't be affected, and those using defaults wouldn't see any shifting anymore.
Again, I'm not sure about enabling `update_in_insert` by default and for me it is a separate discussion with its own performance implications, but fixing the shifts it causes would open the door for at least considering that.

> Could also consider changing the sign text to an empty string and only have a `numhl` with Error -> ErrorMsg, Warn -> WarningMsg, Info/Hint -> Underlined
> 
> More compact, avoids layout shifts and still gives an indication.

That's a good alternative if 'number' is enabled (which is off by default as well). I guess it should be conditionally disabled if signs are shown? It seems like the default behavior just needs to be more adaptive to user options rather than forcing an indication on everyone.

Another thing I noticed: setting `numhl` doesn't forcibly show the number column if the option is disabled, unlike signs. This makes sense because the default 'signcolumn' is 'auto', but are there cases where it is useful? For diagnostics, it seems like the signs provider should decide anyway to avoid interfering with the display. In that case, it feels counterintuitive that the option meant to "show the column only when there's a sign to display" would actually prevent it from displaying in the first place. Perhaps the default 'signcolumn' should be 'no', similar to 'number', and 'auto' reserved for users who want that behavior.

**Comment by justinmk** [comment]
> I'm not in favor of making the "stale" diagnostics behavior the default ([#38313](https://github.com/neovim/neovim/pull/38313)). If a choice must be made, having up-to-date diagnostics on screen is more important than having a signs time to time.

It's not either-or. We can start with #38313 

I'm not really excited about having to enable 'number', or permanent signcolumn, to see diagnostic signs.

**Comment by arkriny** [comment]
> I'm not really excited about having to enable 'number', or permanent signcolumn, to see diagnostic signs.

Then the signs would cause content shift. If it is not the problem, I don't understand why #38313 is needed. Or if it is, why it shouldn't be resolved for all kind of edits, not only for insert mode.

**Comment by justinmk** [comment]
ok +1 for @mfussenegger 's suggestion

**Comment by clason** [comment]
>  still gives an indication.

Unless you don't have `number` set (which is the default!). I'd consider this a rather nasty UX regression.

**Comment by justinmk** [comment]
I also don't use 'number' but was assuming that the underlines were enough? Seems somewhat reasonable to let users opt-in by setting 'number' or signcolumn?

**Comment by clason** [comment]
Depends on the server; the underlines can be very easy to miss if they are not covering a sizable range. And in a greenfield situation, that could be arguable and documented, but given that we already have a default, I think we need a better justification for regressing on that (and I claim that this _is_ a regression, not merely a cosmetic change). Setting `number` is also not an acceptable workaround (if you don't enable them otherwise). I'm not happy about having to force-enable the sign-column in an `LspAttach` autocommand regardless of diagnostics, either (but if that's what I'll have to do, I'll have to do it).

(Also, the diagnostics are not limited to LSP! So it's even harder to argue that non-LSP users should pay for badly behaved servers...)

**Comment by clason** [comment]
I still think some debouncing is the better solution here.

**Comment by arkriny** [comment]
> Setting `number` is also not an acceptable workaround (if you don't enable them otherwise).
> I'm not happy about having to force-enable the sign-column in an `LspAttach` autocommand regardless of diagnostics, either (but if that's what I'll have to do, I'll have to do it).

You don't have to, you can set the old values for `signs` in `vim.diagnostic.config` and have the old behavior. The problem with that is it will require passing the entire `text` table which is too verbose, but there could be a `true` value that maps to the default config, similar to `signs` option.

> I still think some debouncing is the better solution here.

I don't see how debouncing would solve this specific issue of signs disappearing on edit, it would just postpone it as well as creating more complexity and delay.

**Comment by clason** [comment]
> I don't see how debouncing would solve this specific issue of signs disappearing on edit, it would just postpone it as well as creating more complexity and delay.

It would solve it because the issue is not the signs disappearing but the signs _column_ if the signs are (temporarily) gone during refresh.

**Comment by arkriny** [comment]
> (Also, the diagnostics are not limited to LSP! So it's even harder to argue that non-LSP users should pay for badly behaved servers...)

> It would solve it because the issue is not the signs disappearing but the signs _column_ if the signs are (temporarily) gone during refresh.

These two arguments imply the issue is caused by some "badly behaving" servers that refresh diagnostics and cause flickering, which could be fixed with debouncing. But it is a separate issue: any diagnostic display can cause flickering, and a reappearing signcolumn only amplifies this effect.

I think there's a misunderstanding of the original problem. It has been described many times with concrete examples, but to me it boils down to this parity: either the signcolumn for a buffer is always on if signs are ever shown, or signs and the column are never shown. Both approaches avoid jittering; one costs an always-present signcolumn, the other costs missing signs.

For example, most distros choose the first option because they prioritize a richer out of the box experience: [kickstart.nvim](https://github.com/nvim-lua/kickstart.nvim/blob/0619d89884dec2fa911430fdc42796420421fd6b/init.lua#L130), [lazyvim](https://github.com/LazyVim/LazyVim/blob/96f4f18d7d81c786ac0df5723bc7aca058bf2165/lua/lazyvim/config/options.lua#L98) - even has a comment about this, [MiniMax](https://github.com/nvim-mini/MiniMax/blob/c7b5385712531bc1838e1e881edb402e6fbfb259/configs/nvim-0.12/plugin/10_options.lua#L48), [nvchad](https://github.com/NvChad/NvChad/blob/f437558f23c8f50c36cd09748121ab2c822e8ec9/lua/nvchad/options.lua#L34), [AstroNvim](https://github.com/AstroNvim/AstroNvim/blob/7fd58328e2bc91d5cad606ee48fdf924fca6ea3e/lua/astronvim/plugins/_astrocore_options.lua#L38). There are also multiple threads discussing this behavior: [1](https://www.reddit.com/r/neovim/comments/zeanxr/anyone_here_uses_the_option_signcolumn_with_the/), [2](https://www.reddit.com/r/neovim/comments/1idmgl8/my_way_to_avoid_signcolumn_jitter_without/).

This doesn't mean enabling 'signcolumn' is necessarily the right solution for core, but it does demonstrate that there is a problem it affects many users.
